### PR TITLE
Storage: Use rsync when refreshing filesystem volumes in volume-only mode

### PR DESCRIPTION
**Pull request by monstermunchkin**
Fixes https://github.com/lxc/lxd/issues/11308

### CODE REVIEW

File: lxd/storage/backend_lxd.go [hunk 1]
```diff
@@ -149,8 +149,8 @@ func (b *lxdBackend) Driver() drivers.Driver {
 
 // MigrationTypes returns the migration transport method preferred when sending a migration,
 // based on the migration method requested by the driver's ability.
-func (b *lxdBackend) MigrationTypes(contentType drivers.ContentType, refresh bool) []migration.Type {
-	return b.driver.MigrationTypes(contentType, refresh)
+func (b *lxdBackend) MigrationTypes(contentType drivers.ContentType, refresh bool, snapshots bool) []migration.Type {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:152]: Please can you update the docblock here explaining what `snapshots` arg indicates and what its used for.

File: lxd/storage/backend_lxd.go [hunk 2]
```diff
@@ -1386,12 +1386,16 @@ func (b *lxdBackend) RefreshCustomVolume(projectName string, srcProjectName stri
...
-// destination if snapshots are included in the synchronisation.
+// destination if snapshots are included in the synchronisation. An empty srcSnapshots argument
+// indicates a volume-only refresh.
 func (b *lxdBackend) RefreshInstance(inst instance.Instance, src instance.Instance, srcSnapshots []instance.Instance, allowInconsistent bool, op *operations.Operation) error {
 	l := logger.AddContext(b.logger, logger.Ctx{"project": inst.Project().Name, "instance": inst.Name(), "src": src.Name(), "srcSnapshots": len(srcSnapshots)})
 	l.Debug("RefreshInstance started")
 	defer l.Debug("RefreshInstance finished")
 
+	// This indicates whether or not it's a volume-only refresh.
+	snapshots := len(srcSnapshots) > 0
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:1397]: Do we need to differentiate between srcSnapshots being nil or empty?
Is there a difference for optimized refresh between there being no snapshots (but not in instance only mode) vs instance only mode?

Comment (monstermunchkin) [reply lxd/storage/backend_lxd.go:1397]: We don't need to differentiate, I checked that.

Comment (tomponline) [reply lxd/storage/backend_lxd.go:1399]: Thanks.

**Review by tomponline — APPROVED** [review]
LGTM!

**Review by ru-fu — APPROVED** [review]
Docs look good. :)

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
Lets try and get a clear run on ZFS and BTRFS tests.

**Comment by stgraber** [comment]
https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/8887/

still not looking good

**Comment by tomponline** [comment]
Yeah looks like zfs not happy with this one @monstermunchkin

**Comment by monstermunchkin** [comment]
@tomponline I just ran the reproducer with zfs, and it turns out that the issue doesn't exist here. That means we can use optimized refresh on zfs.

**Comment by tomponline** [comment]
> @tomponline I just ran the reproducer with zfs, and it turns out that the issue doesn't exist here. That means we can use optimized refresh on zfs.

How does it work on ZFS without doing a full copy?

**Comment by monstermunchkin** [comment]
@tomponline I guess it just works in some cases. 

This works (last call is significantly faster than the previous ones):

```
 + /home/thomas/go/bin/lxc launch images:ubuntu/jammy c1 -s zfs1
Creating c1
Starting c1
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2

real    0m1.674s
user    0m0.007s
sys     0m0.024s
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2

real    0m1.755s
user    0m0.024s
sys     0m0.006s
+ /home/thomas/go/bin/lxc snapshot c1
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2

real    0m1.674s
user    0m0.007s
sys     0m0.021s
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2 --instance-only

real    0m0.497s
user    0m0.026s
sys     0m0.005s
```

This doesn't:
```
+ /home/thomas/go/bin/lxc launch images:ubuntu/jammy c1 -s zfs1
Creating c1
Starting c1
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2

real    0m1.558s
user    0m0.009s
sys     0m0.022s
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2

real    0m1.452s
user    0m0.014s
sys     0m0.021s
+ /home/thomas/go/bin/lxc snapshot c1
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2 --instance-only

real    0m1.775s
user    0m0.037s
sys     0m0.010s
+ /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2 --instance-only

real    0m1.355s
user    0m0.007s
sys     0m0.021s
```

**Comment by tomponline** [comment]
> @tomponline I guess it just works in some cases.
> 
> This works (last call is significantly faster than the previous ones):
> 
> ```
>  + /home/thomas/go/bin/lxc launch images:ubuntu/jammy c1 -s zfs1
> Creating c1
> Starting c1
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2
> 
> real    0m1.674s
> user    0m0.007s
> sys     0m0.024s
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2
> 
> real    0m1.755s
> user    0m0.024s
> sys     0m0.006s
> + /home/thomas/go/bin/lxc snapshot c1
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2
> 
> real    0m1.674s
> user    0m0.007s
> sys     0m0.021s
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2 --instance-only
> 
> real    0m0.497s
> user    0m0.026s
> sys     0m0.005s
> ```
> 
> This doesn't:
> 
> ```
> + /home/thomas/go/bin/lxc launch images:ubuntu/jammy c1 -s zfs1
> Creating c1
> Starting c1
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2
> 
> real    0m1.558s
> user    0m0.009s
> sys     0m0.022s
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2
> 
> real    0m1.452s
> user    0m0.014s
> sys     0m0.021s
> + /home/thomas/go/bin/lxc snapshot c1
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2 --instance-only
> 
> real    0m1.775s
> user    0m0.037s
> sys     0m0.010s
> + /home/thomas/go/bin/lxc copy c1 c2 --refresh -s zfs2 --instance-only
> 
> real    0m1.355s
> user    0m0.007s
> sys     0m0.021s
> ```

I think for consistency of behavior/documentation we should still fallback to rsync rather than sending the whole volume by ZFS.

**Review by tomponline — CHANGES_REQUESTED** [review]
Lets stick with falling back to rsync for ZFS too for consistency.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 3]
```diff
@@ -2303,22 +2303,17 @@ func (d *zfs) readonlySnapshot(vol Volume) (string, revert.Hook, error) {
...
-		if current != "dev" {
-			// Activate.
-			err = d.setDatasetProperties(dataset, "volmode=dev", "snapdev=visible")
+		if snapdevMode != "visible" {
+			err = d.setDatasetProperties(dataset, "snapdev=visible")
 			if err != nil {
 				return "", nil, err
 			}
 
-			defer func() { _ = d.setDatasetProperties(dataset, "volmode=none", "snapdev=hidden") }()
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2320]: Should this defer not still be here to undo the changes ?

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 4]
```diff
@@ -2332,6 +2332,14 @@ func (d *zfs) readonlySnapshot(vol Volume) (string, revert.Hook, error) {
 
 		mountFlags, mountOptions := filesystem.ResolveMountOptions(strings.Split(vol.ConfigBlockMountOptions(), ","))
 
+		// ext4 will replay the journal if the filesystem is dirty.
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2330]: Ive not seen this in other drivers (I could be wrong) bit I'm interested why this is specifically needed on zfs block vols?

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 3 above)
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go:2320]: Fixed.

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 4 above)
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go:2332]: I have no idea TBH. I ran into issues when trying to mount the ro snapshot, and the logs suggested adding `noload`.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 5]
```diff
@@ -2303,21 +2303,18 @@ func (d *zfs) readonlySnapshot(vol Volume) (string, revert.Hook, error) {
 	d.logger.Debug("Created backup snapshot", logger.Ctx{"dev": srcSnapshot})
 
 	if d.isBlockBacked(vol) {
-		// For block devices, we make them appear.
-		// Check if already active.
-		current, err := d.getDatasetProperty(dataset, "volmode")
+		snapdevMode, err := d.getDatasetProperty(dataset, "snapdev")
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2306]: I'd like to understand the why behind this change? Previously it had a comment which seemed to make sense, so im wondering what the issue was here?

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 4 above)
Comment (tomponline) [reply lxd/storage/drivers/driver_zfs_volumes.go:2332]: It does make some sense. Where were you seeing that error? Have you tried other filesystems too?

I think its not an issue perhaps with lvm as those snapshots are writable, although we could probably do with doing the same thing on all of the block backed filesystem volumes.

Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go:2348]: > It does make some sense. Where were you seeing that error? Have you tried other filesystems too?

I saw the error when `readonlySnapshot()` calls `TryMount()`. I haven't tried other filesystems yet.

Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go:2348]: @tomponline I just tried migration with xfs and btrfs, and both fail at a different stage. They seem unrelated to the rsync change though. Need to fix them.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 6]
```diff
@@ -2835,14 +2844,25 @@ func (d *zfs) MountVolumeSnapshot(snapVol Volume, op *operations.Operation) erro
 					d.logger.Debug("Regenerating filesystem UUID", logger.Ctx{"dev": volPath, "fs": tmpVolFsType})
 					err = regenerateFilesystemUUID(mountVol.ConfigBlockFilesystem(), volPath)
 					if err != nil {
-						return err
+						return nil, err
+					}
+				}
+			} else {
+				// ext4 will replay the journal if the filesystem is dirty.
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2810]: This comment and logic is repeated twice in this PR. It feels like we could do with a driver level function called something like `readOnlyMountOpts(fsType string)` where you can encapsulate that logic and comment once.

Comment (tomponline) [reply lxd/storage/drivers/driver_zfs_volumes.go:2810]: Also, if you specify the same option twice, does it cause any issues? I'm wondering if we need to have the de-dupe logic at all.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 7]
```diff
@@ -2851,14 +2871,25 @@ func (d *zfs) MountVolumeSnapshot(snapVol Volume, op *operations.Operation) erro
...
+			return
+		}
+
+		d.logger.Debug("Unmounted ZFS snapshot dataset", logger.Ctx{"dev": snapshotDataset, "path": mountPath})
+	})
+
+	cleanup := revert.Clone().Fail
 	revert.Success()
-	return nil
+	return cleanup, nil
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2851]: Why the need to return the cleanup hook here rather than calling `UnmountVolumeSnapshot()` when done?

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 8]
```diff
@@ -2341,101 +2341,37 @@ func (d *zfs) readonlySnapshot(vol Volume) (string, revert.Hook, error) {
 		return "", nil, err
 	}
 
-	dataset := d.dataset(vol, false)
-	srcSnapshot := fmt.Sprintf("%s@temp_ro-%s", dataset, uuid.New())
+	snapshotOnlyName := fmt.Sprintf("temp_ro-%s", uuid.New())
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2367]: It feels like `readOnlySnapshot` should return the new temporary Volume that represents the read-only temporary snapshot, that would then allow using `UnmountVolumeSnapshot(vol)` on it, without the need for the new internal `unmountVolumeSnapshot()`. What do you think?

Comment (tomponline) [reply lxd/storage/drivers/driver_zfs_volumes.go:2367]: Also, as the Volume has `SetTemporary` enabled on it, for clarity this function should be called something like `temporaryReadonlySnapshot()`.

**Review by tomponline — CHANGES_REQUESTED** [review]
Please can you add accompanying tests for the ZFS block mode fixes that haven't been detected currently, in order to catch any regressions. Thanks

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 7 above)
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go]: When calling `readonlySnapshot` the cleanup should be done by the caller. That's the way it was, and I didn't change that.

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 8 above)
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go]: I've dropped the whole `SetTemporary` volume thing in favor of an additional argument in `mountVolumeSnapshot`.

**Comment by monstermunchkin** [comment]
Closing this in favor of https://github.com/lxc/lxd/pull/11413.
